File: src/ManagementSystem.cpp
#include "ManagementSystem.h"

ManagementSystem::ManagementSystem()
{
	exitProgram = false;
	databasePersister = new DatabasePersister();
	database = NULL;
	databaseNames = NULL;

	// initially, create the executer with default constructor. This allows the user
	// to run database create / drop commands before specifying a database to use
	executer = new Executer();

	parser = new Parser();

	// load list of database names
	if(!loadDatabaseList())
		cout << "Warning - failed to laod database list. Program may not run as intended";
}


ManagementSystem::~ManagementSystem()
{
	// check if a database object was already loaded, and if save it to the filesystem
	if (database != NULL)
	{
		// databasePersister->saveDatabase((*database));

		delete database;
		delete executer;
	}

	delete databasePersister;
	delete parser;

	// this is redundant, but necessary in case it was forgotten elsewhere
	saveDatabaseList();
}


// reads from an sql file, running one command at a time, until the end of the file
void ManagementSystem::RunInScriptMode(string sqlFilename)
{
	vector<string> *commands = new vector<string>();

	// exit program if it fails to load the file
	if (!getCommandsFromFile(sqlFilename, *commands))
	{
		cout << "Error - file does not exist: " << sqlFilename << endl;
		return;
	}

	// run one command at a time from the file
	for (vector<string>::iterator it = commands->begin(); it != commands->end(); ++it)
	{
    	cout << " > " << *it << endl;
    	processCommand(*it);
 	}
}


// runs the program indefinitely, running a command at a time
void ManagementSystem::RunInCommandLineMode()
{
	string userInput;

	// continually loop while user does not enter "exit" or ".EXIT"
	do
	{		
		// prompt user for input
		cout << " > ";
		getline(cin, userInput, '\n');

		processCommand(userInput);
	} 
	while (!exitProgram);
}


/*
 *	Wrapper function for calling DatabasePersister to try to load database.
 *
 *	Replaces the existing database and executer objects, if they exist.
 * 	Saves the existing database to the file system before replacing it.
 */
bool ManagementSystem::loadDatabase(string dbName)
{
	Database *newDb;
	databasePersister->loadDatabase(dbName, newDb);

	// check if the database actually exists, and return error if it does not
	if (newDb == NULL)
		return false;

	// check if a database object was already loaded, and if save it to the filesystem
	// 	(also replace the old database and executer objects)
	if (database != NULL)
	{
		// databasePersister->saveDatabase((*database));

		delete database;
		delete executer;
	}

	database = newDb;
	executer = new Executer(*database);
	
	return true;
}


// process one SQL command
void ManagementSystem::processCommand(string command)
{
	string lowercaseUserInput;	
	char *token;

	// command to lowercase
	lowercaseUserInput = command;
	transform(lowercaseUserInput.begin(), lowercaseUserInput.end(), lowercaseUserInput.begin(), ::tolower);

	// check for an exit command
	if (lowercaseUserInput.find(".exit") != string::npos || 
		lowercaseUserInput.find("exit") != string::npos)
	{
		cout << "All done." << endl;
		exitProgram = true;
		return;
	}

	// check for a 'use database' command
	else if (lowercaseUserInput.find("use") != string::npos)
	{			
		char *temp = new char[lowercaseUserInput.length() + 1];
		strcpy(temp, lowercaseUserInput.c_str());

		token = strtok(temp, " "); // grabs the 'use' word
		token = strtok(NULL, " "); // ...and now we've grabbed the database name

		if(!loadDatabase(token))
			cout << "!Failed to load '" << token << "' database.";
		else
			cout << "Using database " << token << ".";
	}

	// check for a 'create database' or 'drop database' command
	else if (lowercaseUserInput.find("database") != string::npos)
	{
		cout << processDatabaseCommand(lowercaseUserInput);
	}

	// otherwise, send command to Executer and let it handle the command
	else
	{
		if(executer == NULL)
			cout << "!Failed - must specify a database before accepting any commands.";

		// pass the original command to the executer
		cout << executer->ExecuteCommand(command);
	}

	cout << endl;
}


// processes commands to create and drop databases
string ManagementSystem::processDatabaseCommand(string lowercaseCommand)
{
	string dbName;
	vector<string> commandVector;
	commandVector = parser->SplitCommand(lowercaseCommand);

	// check for invalid command syntax
	if(commandVector.size() != 3)
		return "!Failed to process command, invalid syntax.";

	dbName = commandVector[2];

	// try to create database
	if (commandVector[0] == "create" && commandVector[1] == "database")
	{
		// see if database already exists
		if (databaseExists(dbName))
			return "!Failed to create database " + dbName + " because it already exists.";

		databaseNames->push_back(dbName);
		databasePersister->InitializeDatabase(dbName);
		saveDatabaseList();
		return "Database " + dbName + " created.";
	}

	// try to drop database
	else if (commandVector[0] == "drop" && commandVector[1] == "database")
	{
		// see if database does not exist
		if (!databaseExists(dbName))
			return "!Failed to delete " + dbName + " because it does not exist.";

		databaseNames->remove(dbName);
		databasePersister->DropDatabase(dbName);
		saveDatabaseList();
		return "Database " + dbName + " deleted.";
	}

	else
		return "!Failed to process command, invalid syntax.";
}


// gets all commands from specified SQL file
bool ManagementSystem::getCommandsFromFile(string filename, vector<string> &commands)
{
	ifstream fin;
	string linedata;
	bool readTheFile = false;

	// clear input file-stream flags and open the file
	fin.clear();
	fin.open(filename);

	// read all commands from file into a vector, looking out for comments in the file
	while (fin.good())
	{
		readTheFile = true;

		// get one line at a time
		getline(fin, linedata, '\n');

		// get everything before the first comment, if one exists on the current line
		if(linedata.length() != 0)
		{
			size_t index = linedata.find("--");

			// comment at first character on line, skip this entire line
			if(index == 0)
				continue;

			// no comment was found, add the entire line to vector of commands
			if(index == string::npos)
				commands.push_back(linedata);

			// otherwise, comment was found in middle of line so grab the entire line from before the comment
			else
				commands.push_back(linedata.substr(0, index));
		}
	}

	fin.close();

	return readTheFile;
}


// check if the database exists in the databaseList
bool ManagementSystem::databaseExists(string dbName)
{
	list<string>::iterator it;
	for (it = databaseNames->begin(); it != databaseNames->end(); ++it)
	{

//TEMP
cout << "comparing '" << *it << "' against user specified " << dbName << endl;

    	if ((*it).compare(dbName) == 0)
    		return true;
	}

	return false;
}


// loads the list of database names on file
bool ManagementSystem::loadDatabaseList()
{	
	ifstream fin;
	string linedata;
	bool readTheFile = false;
	string path = "db/databases.meta";

	// initialize database list or return the list if it has already been initialized
	if(databaseNames != NULL)
		return true;
	databaseNames = new list<string>();

	// clear input file-stream flags and open the file
	fin.clear();
	fin.open(path);

	// read all database names into a list
	while (fin.good())
	{
		readTheFile = true;

		// get one line at a time
		getline(fin, linedata, '\n');

		if (linedata.length() > 0)
			databaseNames->push_back(linedata);
	}

	fin.close();

	return readTheFile;
}


// saves the list of database names to file
bool ManagementSystem::saveDatabaseList()
{	
	ofstream fout;
	string linedata;
	bool wroteTheFile = false;
	string path = "db/databases.meta";
	list<string>::iterator it;
	
	it = databaseNames->begin();

	// clear output file-stream flags and open the file
	fout.clear();
  	fout.open(path);

	// read all database names into a list
	while (fout.good() && it != databaseNames->end())
	{
		wroteTheFile = true;

		fout << *it << endl;
		it++;
	}

	fout.close();

	return wroteTheFile;
}

File: java/dbms/src/dbms/ManagementSystem.java
package dbms;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class ManagementSystem {
	
	private static final String DIR = "../db";
	private String workingDirectory = null;
	private String architecture = null;
  	DatabasePersister databasePersister = new DatabasePersister();
  	Database database = null;
    Executer executer = new Executer();
    Parser parser = new Parser();    
    List<String> databaseNames = new ArrayList<String>();
    boolean exitProgram = false;
    
	ManagementSystem(String wdir, String arch) {
		this.workingDirectory = wdir;
		this.architecture = arch;
		
		// attempt to cache database names, and kill the program immediately if this fails
		if (!cacheDatabaseNames())
			exitProgram = true;
	}

	// reads from an sql file, running one command at a time, until the end of the file
	public void RunInScriptMode(String sqlFilename)	{
		List<String> commands = new ArrayList<String>();

		// exit the program if some pre-condition has caused a faulty state
		if(exitProgram) {
			System.out.println("Terminating program due to an error, likely from failure to cache database names.");
			return;
		}
		
		// exit program if it fails to load the file
		if (!getCommandsFromFile(buildSqlFilenamePath(sqlFilename), commands))
		{
			System.out.println(String.format("Error - file does not exist: %1$s.", sqlFilename));
			return;
		}
		else {
			// run one command at a time from the file
			for (String command : commands)
				processCommand(command);
		}
	}


	// runs the program indefinitely, running a command at a time
	public void RunInCommandLineMode() {
		Scanner reader = new Scanner(System.in);
		String command = "";
		String userInput;
		int semicolonIndex;

		// exit the program if some pre-condition has caused a faulty state
		if(exitProgram) {
			System.out.println("Terminating program due to an error, likely from failure to cache database names.");
			reader.close();
			return;
		}
		
		// continually loop while user does not enter "exit" or ".exit"
		do {		
			
			System.out.print(" > ");
			userInput = reader.nextLine();
			
			// check for a .EXIT command
			if (userInput.toLowerCase().equals(".exit")) {
				processCommand(userInput);
			}
			
			// check for a semicolon, and truncate the command at that point if there is one
			semicolonIndex = userInput.indexOf(";");
			
			// there is a semicolon, so build the rest of the command and process it
			if (semicolonIndex != -1) {
				command += userInput.substring(0, semicolonIndex);
				processCommand(command + ";");
				command = "";
				
			}			
			// else there is no semicolon, just append the user input to the working command
			else {
				command += userInput;
			}
			
		} while (!exitProgram);
		
		reader.close();
	}


	/*
	 *	Attempt to load in a new database with the user's specified database name.
	 *	If the specified database exists but a database is already loaded, saves
	 *	the existing database to file before overriding it with the new database. 
	 */
	private boolean loadDatabase(String dbName)
	{
		Database newDb;

		newDb = databasePersister.loadDatabase(buildCompleteDatabasePath(dbName), dbName);

		// check if the database actually exists, and return error if it does not
		if (newDb == null)
			return false;

		// check if a database object was already loaded, and if so save it now
		if (database != null)
			 databasePersister.saveDatabase(buildCompleteDatabasePath(dbName), database);

		database = newDb;
		executer = new Executer(database, buildCompleteDatabasePath(dbName));		
		return true;
	}


	// process one SQL command
	private void processCommand(String command)
	{
		String lowercaseCommand = command.toLowerCase();	

		// check for an exit command
		if (lowercaseCommand.equals(".exit") || lowercaseCommand.equals("exit")) {

			// check if a database object was already loaded, and if so save it now
			if (database != null)
				 databasePersister.saveDatabase(buildCompleteDatabasePath(database.getDatabaseName()), database);
			
			System.out.println("All done.");
			exitProgram = true;
			return;
		}

		// check for a 'use database' command
		else if (lowercaseCommand.startsWith("use ")) {
			String[] tokens = lowercaseCommand.replace(';', ' ').split(" ");
			// load a new database, creating it or fetching it as necessary, and save the existing database
			if (loadDatabase(tokens[1]))
				System.out.println(String.format("Using database %1$s.", tokens[1]));
			else
				System.out.println(String.format("!Failed to load '%1$s' database.", tokens[1]));				
		}

		// check for a 'create database' or 'drop database' command
		else if (lowercaseCommand.startsWith("create database") || lowercaseCommand.startsWith("drop database")) {
			System.out.println(processDatabaseCommand(lowercaseCommand));
		}

		// otherwise, send command to Executer and let it handle the command
		else {
			if (executer == null)
				System.out.println("!Failed - must specify a database before accepting any commands.");
			// else, pass the original command to the executer
			else			
				System.out.println(executer.executeCommand(command));
		}
	}


	// processes commands to create and drop databases
	private String processDatabaseCommand(String lowercaseCommand)
	{
		String dbName;
		List<String> commandTokens = new ArrayList<String>();
		
		for (String token : lowercaseCommand.split(" "))
			commandTokens.add(token);
		
		// check for invalid command syntax
		if(commandTokens.size() != 3)
			return "!Failed to process command, invalid syntax.";

		dbName = commandTokens.get(2).replace(";", "");

		// try to create database
		if (commandTokens.get(0).equals("create") && commandTokens.get(1).equals("database"))
		{
			// see if database already exists
			if (databaseNames.contains(dbName))
				return "!Failed to create database " + dbName + " because it already exists.";

			// otherwise, create the database
			databaseNames.add(dbName);
			databasePersister.initializeDatabase(buildDatabasePath(dbName));
			return "Database " + dbName + " created.";
		}

		// try to drop database
		else if (commandTokens.get(0).equals("drop") && commandTokens.get(1).equals("database"))
		{
			// see if database does not exist
			if (!databaseNames.contains(dbName))
				return "!Failed to delete " + dbName + " because it does not exist.";

			databaseNames.remove(dbName);
			databasePersister.dropDatabase(buildDatabasePath(dbName));
			return "Database " + dbName + " deleted.";
		}

		else
			return "!Failed to process command, invalid syntax.";
	}


	// gets all commands from specified SQL file
	private boolean getCommandsFromFile(String filename, List<String> commands)
	{		
		BufferedReader br = null;
		FileReader fr = null;
		String command = "";
		String linedata;
		int commentIndex = 0;
		int semicolonIndex = 0;

		try {
			fr = new FileReader(filename);
			br = new BufferedReader(fr);
			
			// read a line at a time, watching out for comments
			while ((linedata = br.readLine()) != null) {
				commentIndex = linedata.indexOf("--");
				
				// if there is a comment, only grab the line up to before the comment
				if (commentIndex != -1)
					linedata = linedata.substring(0, commentIndex);
				
				// check if there is a semicolon, now that we have the actual complete command on this line
				semicolonIndex = linedata.trim().indexOf(";");
				
				// if there is a semicolon, grab the command up to that point and append it to the working command
				// NOTE: anything after the (first) semicolon is tossed out 
				if (semicolonIndex != -1) {
					command += linedata.substring(0, semicolonIndex);
					commands.add(command + ";");
					command = "";
				}
				// else if no semicolon, add the linedata to the working command
				else if (linedata.length() > 0) {
					
					// check for .EXIT commands
					if (linedata.toLowerCase().equals(".exit")) {
						commands.add(linedata);
						command = "";
					}
					else
						command += linedata;				
				}
			}
			
		}
		catch (IOException e) {
			e.printStackTrace();
		}
		finally {
			try {
				if (br != null)
					br.close();
				
				if (fr != null)
					fr.close();
			}
			catch (IOException ex) {
				ex.printStackTrace();
			}
		}

		if (commands.size() > 0)
			return true;
		return false;
	}

	
	/*
	 * Cache the list of possible database names, so the program can know if a 
	 * database exists without having to attempt to read from the file system
	 */
	private boolean cacheDatabaseNames() {
		try {
			File dbFolder = new File(buildDatabaseFolderPath(workingDirectory));
			
			if (dbFolder.exists()) {
				File[] listOfFiles = dbFolder.listFiles();

			    for (int i = 0; i < listOfFiles.length; i++) {
			        if (listOfFiles[i].isDirectory()) {
			            databaseNames.add(listOfFiles[i].getName());
			        }
			    }
			}
			// create the db folder if it doesn't exist
			else {
				dbFolder.mkdir();
			}

		}
		catch (Exception e) {
			System.out.println(String.format("!Fatal error while caching database names."));
			e.printStackTrace();
			return false;
		}
		
		return true;
	}

	
	// builds the complete path to a database (so tables can be accessed), in a platform-independent way
	private String buildCompleteDatabasePath(String dbName) {
		String unixString = "%1$s/db/%2$s/";
		String windowsString = "%1$s\\db\\%2$s\\";		
		return buildPath(unixString, windowsString, dbName);
	}
	
	//builds db folder, in a platform-independant way
	private String buildDatabaseFolderPath(String workingDir) {
		String unixString = "%1$s/db";
		String windowsString = "%1$s\\db";		
		return buildPath(unixString, windowsString, workingDir);
	}
	
	
	// builds the path to a database, in a platform-independent way
	private String buildDatabasePath(String dbName) {
		String unixString = "%1$s/db/%2$s";
		String windowsString = "%1$s\\db\\%2$s";		
		return buildPath(unixString, windowsString, dbName);
	}

	
	// builds the path to the user's specified sql script, in a platform-independent way
	private String buildSqlFilenamePath(String sqlFilename) {
		String unixString = "%1$s/%2$s";
		String windowsString = "%1$s\\%2$s";		
		return buildPath(unixString, windowsString, sqlFilename);
	}
	
	
	/*
	 * Builds the path to a particular filename, given the unix and windows variations of the 
	 * formatted string representing the filepath. Returns the full-built filepath that is correct
	 * given the architecture that was passed in at the time the ManagementSystem was initialized.
	 * 
	 * NOTE: Assumes the inputted unix/windows strings have the replacement regex (%1$s) for inserting
	 * the working directory and the filename.
	 */
	private String buildPath(String unixString, String windowsString, String filename) {
		if (architecture.equals("unix"))
			return String.format(unixString, workingDirectory, filename);
		else if (architecture.equals("windows"))
			return String.format(windowsString, workingDirectory, filename);
		else {
			System.out.println(String.format("Error - Invalid architecture detected in the ManagementSystem: %1$s", architecture));
			
			// default to unix in case of bad architecture
			return String.format(unixString, workingDirectory, filename);
		}
	}
}
